### PR TITLE
Issue220 dynamics module

### PR DESCRIPTION
**Pull request by einola**
The bulk of the dynamics merge

### CODE REVIEW

**Comment by dorchard** [comment]
This PR merge depends on first solving https://github.com/nextsimdg/nextsimdg/issues/275

**Comment by timspainNERSC** [comment]
Cleaning up comment out code 1/n.

With all its executive statements, `dynamics/applications/CMakeLists.txt` does nothing. Since all the source code in `applications` is for creating executables, the entire subdirectory is redundant.

So I deleted it and removed the reference to it from the `dynamics` CMakeLists.txt.

**Comment by einola** [comment]
This pends on #203 and conflict resolution

**Comment by timspainNERSC** [comment]
Conflicts keep popping up like bubbles under wallpaper!

**Comment by einola** [comment]
I'm not a reviewer, but I say it's time to merge.

File: dynamics/src/include/DynamicsKernel.hpp [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+    /*
+        Temrorary Ocen and Wind Field
+    */
+
+    class OceanX : public Nextsim::Interpolations::Function {
+    public:
+        double operator()(double x, double y) const
+        {
+            return  0.01 * (2.0 * y / ( 121. * 25000.) - 1.0);
```
Comment (dorchard) [inline dynamics/src/include/DynamicsKernel.hpp:100]: There are some magic numbers here which are probably better extracted as constants, particular 121/154 and 25000 (here and line 107, 121,122,142,143)

File: run/era5_topaz4_maker.py [hunk 2]
```diff
@@ -65,37 +65,47 @@ def topaz4_source_file_name(field, unix_time):
...
 
+    iwrap = (i + 1) % data.shape[1]
+
     dataplier = data != missing # False is zero
 
     weighted_sum = ((1 - fj) * (1 - fi) * data[j, i] * dataplier[j, i] +
-        (1 - fj) * (fi) * data[j, i + 1] * dataplier[j, i] +
-        (fj) * (1 - fi) * data[j + 1, i] * dataplier[j, i] +
-        (fj) * (fi) * data[j + 1, i + 1] * dataplier[j, i])
+        (1 - fj) * (fi) * data[j, iwrap] * dataplier[j, iwrap] +
```
Comment (dorchard) [inline run/era5_topaz4_maker.py:101]: Are we sure the second `iwrap` here is right? The code it replaces was `i` but this is now `i + 1` modded.

File: run/era5_topaz4_maker.py [hunk 3]
```diff
@@ -65,37 +65,47 @@ def topaz4_source_file_name(field, unix_time):
...
+        (fj) * (1 - fi) * data[j + 1, i] * dataplier[j + 1, i] +
```
Comment (dorchard) [inline run/era5_topaz4_maker.py:102]: Is the second `j+1` right here? The previous code was `j`.

File: run/era5_topaz4_maker.py [hunk 4]
```diff
@@ -65,37 +65,47 @@ def topaz4_source_file_name(field, unix_time):
...
+        (fj) * (fi) * data[j + 1, iwrap] * dataplier[j + 1, iwrap])
```
Comment (dorchard) [inline run/era5_topaz4_maker.py:103]: Is the second `iwrap` correct here? (not `i`? as in the previous code)?

File: run/era5_topaz4_maker.py (hunk 2 above)
Comment (dorchard) [reply run/era5_topaz4_maker.py:101]: Or do we need ` i % data.shape[1]`?

**Review by dorchard — COMMENTED** [review]
I've done a light touch look through things and spotted a few potential problems. They could be nothing though, but thought best to flag.

File: core/src/modules/include/Dynamics.hpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+class Dynamics : public IDynamics {
+public:
+    Dynamics();
+
+    std::string getName() const override { return "Dynamics"; }
+    void update(const TimestepTime& tst) override;
+
+    void setData(const ModelState::DataMap&) override;
+private:
+    // TODO: How to get the template parameters here?
```
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply core/src/modules/include/Dynamics.hpp:28]: #360

File: run/era5_topaz4_maker.py (hunk 2 above)
Comment (timspainNERSC) [reply run/era5_topaz4_maker.py:101]: This is now correct. The indices of `data` and `dataplier` should match. That `dataplier` had [j, i] for all the terms was a copy-paste error. The terms for basic bilinear interpolation should be [j, i], [j, i+1], [j+1, i] and [j+1, i+1]. Here i+1 is replaced by iwrap to account for the wrap-around in longitude.

File: run/era5_topaz4_maker.py (hunk 3 above)
Comment (timspainNERSC) [reply run/era5_topaz4_maker.py:102]: See comment on line 101.

File: run/era5_topaz4_maker.py (hunk 4 above)
Comment (timspainNERSC) [reply run/era5_topaz4_maker.py:103]: See comment on line 101.

File: dynamics/src/DGTransport.cpp [hunk 6]
```diff
@@ -183,11 +182,31 @@ void DGTransport<DG>::reinitnormalvelocity()
...
+    // the normal is set only once, from the inside. These edges must be scaled with 2.0
+    
+    for (size_t seg = 0 ; seg<4; ++seg) // run over the 4 segments (bot, right, top, left)
+      {
+#pragma omp parallel for
+	for (size_t i = 0; i<smesh.dirichlet[seg].size();++i)
+	  {
+	    const size_t eid = smesh.dirichlet[seg][i];  //! The i of the boundary element 
+	    const size_t ix = eid % smesh.nx;            //! x & y indices of the element
+           const size_t iy = eid / smesh.nx;
```
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply dynamics/src/DGTransport.cpp:198]: No, using nx in both lines is correct. This is the usual way of getting 2-dimensional indices from a 1-dimensional index.

File: dynamics/src/include/DynamicsKernel.hpp (hunk 1 above)
Comment (timspainNERSC) [reply dynamics/src/include/DynamicsKernel.hpp:100]: Very much so!
However all of this code is removed [here](https://github.com/nextsimdg/nextsimdg/pull/327/files#diff-906d264d24ee76696155c8cabfaeccccf0fbe08736c467b65e7ebb705f534015) as part of the development that gets the main model and the dynamics talking to each other correctly. After merging this PR, the dynamics still has no effect on the model, so these awful hard coded constants can be ignored for now.

File: dynamics/src/DGTransport.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (dorchard) [reply dynamics/src/DGTransport.cpp:198]: Thanks

File: dynamics/src/include/DynamicsKernel.hpp (hunk 1 above)
Comment (dorchard) [reply dynamics/src/include/DynamicsKernel.hpp:100]: Gotcha.

**Review by dorchard — APPROVED** [review]

**Pull request merged**